src/UI/Component/Input/Field/TagInput.php

```diff
@@ -0,0 +1,134 @@
+<?php
+declare(strict_types=1);
+
+namespace ILIAS\UI\Component\Input\Field;
+
+use ILIAS\UI\Component\JavaScriptBindable;
+use ILIAS\UI\Component\Signal;
+
+/**
+ * Interface TagInput
+ *
+ * This describes Tag Inputs
+ *
+ * @package ILIAS\UI\Component\Input\Field
+ */
+interface TagInput extends Input, JavaScriptBindable {
+
+	const EVENT_ITEM_ADDED = 'itemAdded';
+	const EVENT_BEFORE_ITEM_REMOVE = 'beforeItemRemove';
+	const EVENT_BEFORE_ITEM_ADD = 'beforeItemAdd';
+	const EVENT_ITEM_REMOVED = 'itemRemoved';
+
+
+	/**
+	 * @return array of tags such as [ 'Interesting', 'Boring', 'Animating', 'Repetitious' ]
+	 */
+	public function getTags(): array;
+
+
+	/**
+	 * Get an input like this, but decide whether the user can provide own
+	 * tags or not. (Default: Allowed)
+	 *
+	 * @param bool $extendable
+	 *
+	 * @return \ILIAS\UI\Component\Input\Field\TagInput
+	 */
+	public function withUserCreatedTagsAllowed(bool $extendable): TagInput;
+
+
+	/**
+	 * @see withUserCreatedTagsAllowed
+	 * @return bool Whether the user is allowed to input more
+	 * options than the given.
+	 */
+	public function areUserCreatedTagsAllowed(): bool;
+
+
+	/**
+	 * Get an input like this, but change the amount of characters the
+	 * user has to provide before the suggestions start (Default: 1)
+	 *
+	 * @param int $characters , defaults to 1
+	 *
+	 * @return \ILIAS\UI\Component\Input\Field\TagInput
+	 */
+	public function withSuggestionsStartAfter(int $characters): TagInput;
+
+
+	/**
+	 * @see withSuggestionsStartAfter
+	 * @return int
+	 */
+	public function getSuggestionsStartAfter(): int;
+
+
+	/**
+	 * Get an input like this, but limit the amount of characters one tag can be. (Default: unlimited)
+	 *
+	 * @param int $max_length
+	 *
+	 * @return \ILIAS\UI\Component\Input\Field\TagInput
+	 */
+	public function withTagMaxLength(int $max_length): TagInput;
+
+
+	/**
+	 * @see withTagMaxLength
+	 * @return int
+	 */
+	public function getTagMaxLength(): int;
+
+
+	/**
+	 * Get an input like this, but limit the amount of tags a user can select or provide. (Default: unlimited)
+	 *
+	 * @param int $max_tags
+	 *
+	 * @return \ILIAS\UI\Component\Input\Field\TagInput
+	 */
+	public function withMaxTags(int $max_tags): TagInput;
+
+
+	/**
+	 * @see withMaxTags
+	 * @return int
+	 */
+	public function getMaxTags(): int;
+
+
+	// Events
```

Comment:
I'm still wondering I we really want to expose all these events to the user. What would I want to do with the `onBefore`-events?

Reply:
you could decide whether the tag can be added ore not

Reply:
Ah ok, I see. So you get that `onBefore`-event and abort the further processing of the event in your handler and this then stops the actual adding of that tag to the field. Puuh, intricate stuff.

I definitely see the use case, but I don't like handling it that way. This will introduce a dependency of the behaviour to the order in which one added event handlers (right?)? This doesn't help telling the user, why her tag wasn't added. After all, we will reintroduce the well known problems of javascript event handling (aka "why did that just happen??") in that way.

Let's think of a less error prone way to achieve the desired behaviour.